source/httpfs/driver.go

```diff
@@ -0,0 +1,159 @@
+package httpfs
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path"
+
+	"github.com/golang-migrate/migrate/v4/source"
+)
+
+// driver is a migration source driver for reading migrations from
+// http.FileSystem instances. It implements source.Driver interface and can be
+// used as a migration source for the main migrate library.
+type driver struct {
+	migrations *source.Migrations
+	fs         http.FileSystem
+	path       string
+}
+
+// New creates a new migrate source driver from a http.FileSystem instance and a
+// relative path to migration files within the virtual FS. It is identical to
+// the WithInstance() function except it will delay any errors on fist usage of
+// this driver. This reduces the number of error handling branches in clients of
+// this package without compromising on correctness.
+func New(fs http.FileSystem, path string) source.Driver {
+	d, err := newDriver(fs, path)
+	if err != nil {
+		return &failedDriver{err}
+	}
+	return d
+}
+
+// WithInstance creates new migrate source driver from a http.FileSystem
+// instance and a relative path to migration files within the virtual FS.
+func WithInstance(fs http.FileSystem, path string) (source.Driver, error) {
+	return newDriver(fs, path)
+}
+
+func newDriver(fs http.FileSystem, path string) (*driver, error) {
+	root, err := fs.Open(path)
+	if err != nil {
+		return nil, err
+	}
+
+	files, err := root.Readdir(0)
+	if err != nil {
+		_ = root.Close()
+		return nil, err
+	}
+	if err = root.Close(); err != nil {
+		return nil, err
+	}
+
+	ms := source.NewMigrations()
+	for _, file := range files {
+		if file.IsDir() {
+			continue
+		}
+
+		m, err := source.DefaultParse(file.Name())
+		if err != nil {
+			continue // ignore files that we can't parse
+		}
+
+		if !ms.Append(m) {
+			return nil, fmt.Errorf("duplicate migration file: %v", file.Name())
+		}
+	}
+
+	return &driver{
+		fs:         fs,
+		path:       path,
+		migrations: ms,
+	}, nil
+}
+
+// Open is part of source.Driver interface implementation. It always returns
+// error because http.FileSystem must be provided by the user of this package
+// and created using WithInstance() function.
+func (h *driver) Open(url string) (source.Driver, error) {
+	return nil, fmt.Errorf("not implemented")
+}
+
+// Close is part of source.Driver interface implementation. This is a no-op.
+func (h *driver) Close() error {
+	return nil
+}
+
+// First is part of source.Driver interface implementation.
+func (h *driver) First() (version uint, err error) {
+	if version, ok := h.migrations.First(); ok {
+		return version, nil
+	}
+	return 0, &os.PathError{
+		Op:   "first",
+		Path: h.path,
+		Err:  os.ErrNotExist,
+	}
+}
+
+// Prev is part of source.Driver interface implementation.
+func (h *driver) Prev(version uint) (prevVersion uint, err error) {
+	if version, ok := h.migrations.Prev(version); ok {
+		return version, nil
+	}
+	return 0, &os.PathError{
+		Op:   fmt.Sprintf("prev for version %v", version),
```

Comment:
I personally pretend that go language does not support string concatenation with `+` operator and never use it.

```go
               Op:   fmt.Sprintf("prev for version %v", version),
```
vs
```go
               Op:   "prev for version " + strconv.Itoa(int(version)),
```

Using strconv requires additional uint -> int typecasting. It will produce strange results if version value overflows signed integer range. To make it proper we should use `FormatUint()`. This leaves us with:

```go
               Op:   fmt.Sprintf("prev for version %v", version),
```
vs
```go
               Op:   "prev for version " + strconv.FormatUint(uint64(version), 10),
```

Any chance you would consider leaving `fmt.Sprintf()`?

Reply:
I don't like `fmt` due to the perf hit but like the readability when there's a bunch of formatting to be done. In this case, the formatting is simple so I lean towards string concatenation. Also, any formatting that's done in a loop I'll sacrifice more readability for performance.

Reply:
One could argue that performance hit is not that big. Could not resist trying it out:

```go
func BenchmarkSprintf(b *testing.B) {
        var s string
        for i := 0; i < b.N; i++ {
                s = fmt.Sprintf("prev for version %v", b.N%100)
        }
        _ = s
}

func BenchmarkStringPlustItoa(b *testing.B) {
        var s string
        for i := 0; i < b.N; i++ {
                s = "prev for version " + strconv.Itoa(b.N%100)
        }
        _ = s
}
```

```sh
$ go test -bench . tmp_test.go 
goos: linux
goarch: amd64
BenchmarkSprintf-4           	 7888105	       165 ns/op
BenchmarkStringPlustItoa-4   	18204394	        60.7 ns/o
```

Still you are calling all the shots here. Plus it is important to keep the style consistent. Lets switch to concatenation :+1:.